python/lsst/ts/ATHexapod/controller.py

```diff
@@ -121,22 +122,24 @@ def is_connected(self):
 
             if has_response:
                 try:
-                    keep_going = multi_line
+                    keep_going = False if num_line == 0 else True
                     while True:
-                        c = await asyncio.wait_for(self.reader.read(1), timeout=2.)
-                        response += c
+                        line = await self.reader.readuntil(b'\n')
+                        response += line
+                        num_line -= 1
+                        self.log.debug(f"Response: {response}\n num_line: {num_line}")
 
-                        if c == b'\n' and not keep_going:
+                        keep_going = False if num_line == 0 else True
+                        if not keep_going:
                             break
-                        elif c == b' ':
-                            keep_going = True
-                        elif c == b'\n':
-                            keep_going = False
                 except asyncio.TimeoutError:
                     self.log.warning("Timed out waiting for response from controller. Result "
                                      "may be incomplete.")
 
         self.log.debug(f"Response: {response!r}")
+        response = response.decode("ISO-8859-1")
+        response = response.replace("\n", "")
+        response = response.split(" ")
```

Comment:
Based on some of your test code below I am guessing that the spaces your code deals with are just leading spaces that are meaningless. For example `f"X=1\n Y=1\n Z=1\n U=1\n V=1\n W=1\n"`. If that is the case then please consider using `strip` to get rid of them. I think the following code does everything you need:
```
replies = []
for i in range(num_line):
    raw_line = await self.reader.readuntil(b'\n')
    line = raw_line.decode("ISO-8859-1").strip()  # Remove trailing \n and leading whitespace
    self.log.debug(f"Read {i+1} of {num_line} lines: {line}")
    replies.append(line)
```
This will need work if num_line can be 0, but I didn't find any such calls.

Reply:
I strongly suggest raising an exception if the read times out. That makes the calling code simpler because it can assume it gets the right number of replies. I admit it may require trapping that exception somewhere up calling chain, unless you can just ignore it and log the error in `write_command`.